sched/event/event_post.c

```diff
@@ -0,0 +1,162 @@
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxevent_post
+ *
+ * Description:
+ *   Post one or more events to an event object.
+ *
+ *   This routine posts one or more events to an event object. All tasks
+ *   waiting on the event object event whose waiting conditions become
+ *   met by this posting immediately unpend.
+ *
+ *   Posting differs from setting in that posted events are merged together
+ *   with the current set of events tracked by the event object.
+ *
+ * Input Parameters:
+ *   event  - Address of the event object
+ *   events - Set of events to post to event
+ *          - Set events to 0 will be considered as any,
+ *            waking up the waiting thread immediately.
+ *
+ * Returned Value:
+ *   This is an internal OS interface and should not be used by applications.
+ *   It follows the NuttX internal error return policy:  Zero (OK) is
+ *   returned on success.  A negated errno value is returned on failure.
+ *
+ * Assumptions:
+ *   This function may be called from an interrupt handler.
+ *
+ ****************************************************************************/
+
+int nxevent_post(FAR nxevent_t *event, nxevent_mask_t events,
+                 nxevent_flags_t eflags)
+{
+  nxevent_mask_t clear = 0;
+  FAR nxevent_wait_t *wait;
+  FAR nxevent_wait_t *tmp;
+  irqstate_t flags;
+  bool waitall;
+  bool postall;
+  int ret = 0;
+
+  if (event == NULL)
+    {
+      return -EINVAL;
+    }
+
+  if (events == 0)
+    {
+      events = ~0;
+    }
+
+  flags = enter_critical_section();
+
+  event->events |= events;
+
+  if (!list_is_empty(&event->list))
+    {
+      postall = ((eflags & NXEVENT_POST_ALL) != 0);
+
+      /* Hold schedule lock here to avoid context switch if post high
+       * priority task.
+       */
+
+      sched_lock();
+
+      list_for_every_entry_safe(&event->list, wait, tmp,
+                                nxevent_wait_t, node)
+        {
+          waitall = ((wait->eflags & NXEVENT_WAIT_ALL) != 0);
+
+          if ((!waitall && ((wait->expect & event->events) != 0)) ||
+              (waitall && ((wait->expect & event->events) == wait->expect)))
+            {
+              list_delete(&wait->node);
+
+              if (!waitall)
+                {
+                  wait->expect &= event->events;
+                }
+
+              if (!postall && (wait->eflags & NXEVENT_WAIT_NOCLEAR) == 0)
+                {
+                  clear |= wait->expect;
+                }
+
+              ret = nxevent_sem_post(&wait->sem);
+              if (ret >= 0 && (event->events & ~clear) == 0)
+                {
+                  break;
+                }
+            }
+        }
+
+      sched_unlock();
+
+      if (clear)
+        {
+          event->events &= ~clear;
+        }
+    }
+
+  if ((eflags & NXEVENT_POST_SET) != 0)
```

Comment:
merge to line 106

Reply:
No, here should be initialized to the specified value before exiting

Reply:
should remove thewhole if block now

Reply:
I want to keep the semantics of `nxevent_post(0)`, if NXEVENT_POST_SET is set then the final value should be 0, not ~0